### PR TITLE
Add `EXTRACT` Function To OpenSearch SQL Plugin

### PR DESCRIPTION
**Pull request by GabeFernandez310**
### Description
Adds the `EXTRACT` function to the SQL plugin, aligned with the [MySQL documentation](https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html#function_extract). Extract(_part_ FROM _date_) returns a number formatted according to what is defined by the temporal interval being used for the _part_ argument. A list of the MySQL temporal intervals can be found [here](https://dev.mysql.com/doc/refman/8.0/en/expressions.html#temporal-intervals).

Example:
`Extract(DAY FROM "2023-02-11 10:11:12.123")` -> `11`
`Extract(DAY SECOND_MICROSECOND "2023-02-11 10:11:12.123")` -> `12123000`
 
### Issues Resolved
[722](https://github.com/opensearch-project/sql/issues/722)
 
### Check List
- [X] New functionality includes testing.
  - [X] All tests pass, including unit test, integration test and doctest
- [X] New functionality has been documented.
  - [X] New functionality has javadoc added
  - [X] New functionality has user manual doc added
- [X] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: sql/src/main/antlr/OpenSearchSQLParser.g4 [hunk 1]
```diff
@@ -323,6 +324,32 @@ getFormatType
     | TIMESTAMP
     ;
 
+extractFunction
+    : EXTRACT LR_BRACKET datetimePart FROM functionArg RR_BRACKET
+    ;
+
+datetimePart
+    : MICROSECOND
```
Comment (acarbonetto) [inline sql/src/main/antlr/OpenSearchSQLParser.g4:332]: can you put this alphabetic order?

**Review by acarbonetto — DISMISSED** [review]
LGMT

**Review by matthewryanwells — APPROVED** [review]

**Review by matthewryanwells — APPROVED** [review]

File: core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFormatterUtil.java [hunk 2]
```diff
@@ -62,6 +62,7 @@ interface DateTimeFormatHandler {
       .put("%W", (date) -> "EEEE") // %W => EEEE - Weekday name (Sunday..Saturday)
       .put("%Y", (date) -> "yyyy") // %Y => yyyy - Year, numeric, 4 digits
       .put("%y", (date) -> "yy") // %y => yy - Year, numeric, 2 digits
+      .put("%Q", (date) -> "Q") // %Q => Q - Quarter, numeric, 1 digit
```
Comment (Yury-Fridlyand) [inline core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFormatterUtil.java:65]: This is not needed any more I guess

File: core/src/test/java/org/opensearch/sql/expression/datetime/ExtractTest.java [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+import org.opensearch.sql.data.model.ExprDateValue;
+import org.opensearch.sql.data.model.ExprDatetimeValue;
+import org.opensearch.sql.data.model.ExprTimeValue;
+import org.opensearch.sql.data.model.ExprValue;
+import org.opensearch.sql.expression.DSL;
+import org.opensearch.sql.expression.Expression;
+import org.opensearch.sql.expression.ExpressionTestBase;
+import org.opensearch.sql.expression.FunctionExpression;
+
+@ExtendWith(MockitoExtension.class)
```
Comment (Yury-Fridlyand) [inline core/src/test/java/org/opensearch/sql/expression/datetime/ExtractTest.java:30]: This is not needed

File: core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java [hunk 4]
```diff
@@ -7,6 +7,7 @@
 package org.opensearch.sql.expression.datetime;
 
 import static java.time.temporal.ChronoField.ALIGNED_WEEK_OF_YEAR;
+import static java.time.temporal.ChronoField.YEAR;
```
Comment (Yury-Fridlyand) [inline core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java:10]: Could be reverted

File: sql/src/test/java/org/opensearch/sql/sql/antlr/SQLSyntaxParserTest.java [hunk 5]
```diff
@@ -175,6 +175,37 @@ private static Stream<Arguments> nowLikeFunctionsData() {
     );
   }
 
+  private static Stream<Arguments> getPartForExtractFunction() {
+    return Stream.of(
+        Arguments.of("MICROSECOND"),
```
Comment (Yury-Fridlyand) [inline sql/src/test/java/org/opensearch/sql/sql/antlr/SQLSyntaxParserTest.java:180]: Probably need a UT or IT with incorrect `part` argument:
```sql
select extract(cat from date '1984-04-04')
```

**Review by Yury-Fridlyand — APPROVED** [review]

File: sql/src/test/java/org/opensearch/sql/sql/antlr/SQLSyntaxParserTest.java (hunk 5 above)
Comment (GabeFernandez310) [reply sql/src/test/java/org/opensearch/sql/sql/antlr/SQLSyntaxParserTest.java:180]: Added in https://github.com/opensearch-project/sql/pull/1421/commits/a38212cbef93f40de3e26ff6628fcfdd9b9660c0

File: core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java (hunk 4 above)
Comment (GabeFernandez310) [reply core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java]: Removed in https://github.com/opensearch-project/sql/pull/1421/commits/a38212cbef93f40de3e26ff6628fcfdd9b9660c0

File: core/src/test/java/org/opensearch/sql/expression/datetime/ExtractTest.java (hunk 3 above)
Comment (GabeFernandez310) [reply core/src/test/java/org/opensearch/sql/expression/datetime/ExtractTest.java]: Removed in https://github.com/opensearch-project/sql/pull/1421/commits/a38212cbef93f40de3e26ff6628fcfdd9b9660c0

File: core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFormatterUtil.java (hunk 2 above)
Comment (GabeFernandez310) [reply core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFormatterUtil.java]: Removed in https://github.com/opensearch-project/sql/pull/1421/commits/a38212cbef93f40de3e26ff6628fcfdd9b9660c0

**Review by Yury-Fridlyand — APPROVED** [review]

**Review by dai-chen — APPROVED** [review]

**Pull request merged**